### PR TITLE
18 - GraphQL and Postgres setup, animals, animal queries

### LINKED ISSUE #18: Add first graphql endpoint to fetch some data
Select which data to fetch and show example how to create graphql "endpoint".

### PR DESCRIPTION
**Pull request by psaulenas**
Closes #18

GraphQL and Postgres setup, animals, animal queries

Example queries:

- Animals

<details>
  <summary>Query Example</summary>

  ```graphql
{
  animals {
    id
    organization
    registration_no
    registration_date
    status
    image_url
    birth_date
    name
    species
    breed
    gender
    color
    weight
    microchip_id
    chip_install_date
    allergy
    food
    comments
    mod_time
  }
}
  ```
</details>

<details>
  <summary>Response Example</summary>

  ```json
{
  "data": {
    "animals": [
      {
        "id": 2,
        "organization": "2",
        "registration_no": "321Hi",
        "registration_date": "1608422400000",
        "status": "sick",
        "image_url": "www.url.lt",
        "birth_date": "1577836800000",
        "name": "Jupyteris",
        "species": "parrot",
        "breed": null,
        "gender": "female",
        "color": "green",
        "weight": 1,
        "microchip_id": "666666",
        "chip_install_date": "1598918400000",
        "allergy": null,
        "food": null,
        "comments": null,
        "mod_time": "1608462526888"
      },
      {
        "id": 3,
        "organization": "3",
        "registration_no": "456Carl",
        "registration_date": "1608422400000",
        "status": "healthy",
        "image_url": "www.url.lt",
        "birth_date": "1577836800000",
        "name": "Pitoncas",
        "species": "snake",
        "breed": null,
        "gender": "male",
        "color": "green",
        "weight": 0.3,
        "microchip_id": "2893402",
        "chip_install_date": "1583020800000",
        "allergy": null,
        "food": null,
        "comments": null,
        "mod_time": "1608462526888"
      },
      {
        "id": 4,
        "organization": "3",
        "registration_no": "555Bob",
        "registration_date": "1608422400000",
        "status": "adopted",
        "image_url": "www.url.lt",
        "birth_date": "1577836800000",
        "name": "Javainis",
        "species": "hamster",
        "breed": null,
        "gender": "male",
        "color": "orange",
        "weight": 0.4,
        "microchip_id": "29387",
        "chip_install_date": "1586822400000",
        "allergy": null,
        "food": null,
        "comments": null,
        "mod_time": "1608462526888"
      },
      {
        "id": 5,
        "organization": "1",
        "registration_no": "999Mamba",
        "registration_date": "1608422400000",
        "status": "healthy",
        "image_url": "www.url.lt",
        "birth_date": "1568592000000",
        "name": "Murkė",
        "species": "cat",
        "breed": "ragdoll",
        "gender": "male",
        "color": "grey",
        "weight": 7,
        "microchip_id": "001010101",
        "chip_install_date": "1598918400000",
        "allergy": null,
        "food": null,
        "comments": null,
        "mod_time": "1608462526888"
      },
      {
        "id": 1,
        "organization": "1",
        "registration_no": "123Svx",
        "registration_date": "1608422400000",
        "status": "vaccinated",
        "image_url": "www.green.mamba",
        "birth_date": "1596240000000",
        "name": "Haskelis",
        "species": "dog",
        "breed": "husky",
        "gender": "female",
        "color": "white",
        "weight": 20,
        "microchip_id": "123",
        "chip_install_date": "1597104000000",
        "allergy": null,
        "food": null,
        "comments": null,
        "mod_time": "1608462526890"
      }
    ]
  }
}
  ```
</details>

- Animal

<details>
  <summary>Query Example</summary>

  ```graphql
{
  animal(id: 3) {
    id
    organization
    registration_no
    registration_date
    status
    image_url
    birth_date
    name
    species
    breed
    gender
    color
    weight
    microchip_id
    chip_install_date
    allergy
    food
    comments
    mod_time
  }
}
  ```
</details>

<details>
  <summary>Response Example</summary>

  ```json
{
  "data": {
    "animal": {
      "id": 3,
      "organization": "3",
      "registration_no": "456Carl",
      "registration_date": "1608422400000",
      "status": "healthy",
      "image_url": "www.url.lt",
      "birth_date": "1577836800000",
      "name": "Pitoncas",
      "species": "snake",
      "breed": null,
      "gender": "male",
      "color": "green",
      "weight": 0.3,
      "microchip_id": "2893402",
      "chip_install_date": "1583020800000",
      "allergy": null,
      "food": null,
      "comments": null,
      "mod_time": "1608462526888"
    }
  }
}
  ```
</details>

### CODE REVIEW

File: src/utils/postgres.ts [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+  const query = {
+    text,
+    values: [id],
+  };
+
+  return query;
+};
+
+export const getAnimalsQuery = (): QueryConfig => {
+  const text = `SELECT
```
Comment (ignasposka) [inline src/utils/postgres.ts:36]: Why not `SELECT *`?

**Review by ignasposka — CHANGES_REQUESTED** [review]
Good job in general! But still few things I have noticed.

File: Dockerfile [hunk 2]
```diff
@@ -1,4 +1,4 @@
 FROM node:12
-WORKDIR /usr/src/app
```
[the remark this answers is not in the document]
Comment (psaulenas) [reply Dockerfile:2]: I removed `package.json` and other config files from the `src` folder in order to lint the source code, before this change the code in `node_modules` was also checked. And I personally think that config files shouldn't be in the `src`.

File: package.json [hunk 3]
```diff
@@ -6,19 +6,35 @@
...
+    "pg": "^8.5.1"
   },
   "devDependencies": {
     "@types/express": "^4.17.9",
+    "@types/graphql": "^14.5.0",
+    "@types/lodash": "^4.14.165",
+    "@types/pg": "^7.14.7",
+    "husky": "^4.3.6",
     "nodemon": "^2.0.6",
+    "tslint": "^6.1.3",
```
[the remark this answers is not in the document]
Comment (psaulenas) [reply package.json:35]: I am happy to change `tslint` to `eslint`

File: src/index.ts [hunk 4]
```diff
@@ -1,47 +1,47 @@
...
-  resolvers,
+import express from 'express';
+import bodyParser from 'body-parser';
+import { graphqlExpress, graphiqlExpress } from 'apollo-server-express';
+
+import { schema } from './schemas';
+import initClients from './utils/init-clients';
+
+initClients().then(({ pgClient }) => {
+  // Initialize the app
```
[the remark this answers is not in the document]
Comment (psaulenas) [reply src/index.ts:9]: Happy to remove them, @JustasBelevicius left them 😄

File: src/utils/postgres.ts (hunk 1 above)
Comment (psaulenas) [reply src/utils/postgres.ts:36]: I think we should list all columns explicitly. `SELECT *` makes code hard to maintain (you have to know the structure of the source tables) and also opens up the risk of breakages.

Comment (ignasposka) [reply src/utils/postgres.ts:36]: Alright, let's leave it like this then.

**Review by ignasposka — APPROVED** [review]

File: src/utils/postgres.ts [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+                    color,
+                    weight,
+                    microchip_id,
+                    chip_install_date,
+                    allergy,
+                    food,
+                    comments,
+                    mod_time
+                FROM public.animal
+                WHERE id = $1;`;
```
Comment (AzisK) [inline src/utils/postgres.ts:25]: Is this safe from SQL injections?

**Review by AzisK — APPROVED** [review]
LGTM

File: src/utils/postgres.ts (hunk 5 above)
Comment (psaulenas) [reply src/utils/postgres.ts:25]: Yes, it is parameterized query, `QueryConfig` is returned from this function

**Pull request merged**